src/metabase/query_processor/util/nest_query.clj

```diff
@@ -0,0 +1,113 @@
+  "Convert an `:expression` reference from a source query into an appropriate `:field` clause for use in the surrounding
+  query."
+  [{:keys [expressions source-query], :as query} [_ expression-name opts :as clause]]
+  (let [expression-definition        (or (get expressions (keyword expression-name))
+                                         (throw (ex-info (tru "No expression named {0}" (pr-str expression-name))
+                                                         {:type            qp.error-type/invalid-query
+                                                          :expression-name expression-name
+                                                          :query           query})))
+        {base-type :base_type}       (some-> expression-definition annotate/infer-expression-type)
+        {::add/keys [desired-alias]} (mbql.u/match-one source-query
+                                       [:expression (_ :guard (partial = expression-name)) source-opts]
+                                       source-opts)]
+    [:field
+     (or desired-alias expression-name)
+     (assoc opts :base-type (or base-type :type/*))]))
+
+(defn- rewrite-fields-and-expressions [query]
+  (mbql.u/replace query
+    ;; don't rewrite anything inside any source queries or source metadata.
+    (_ :guard (constantly (some (partial contains? (set &parents))
+                                [:source-query :source-metadata])))
+    &match
+
+    :expression
+    (raise-source-query-expression-ref query &match)
+
+    ;; mark all Fields at the new top level as `::outer-select` so QP implementations know not to apply coercion or
+    ;; whatever to them a second time.
+    [:field _id-or-name (_opts :guard (every-pred :temporal-unit (complement ::outer-select)))]
+    (recur (mbql.u/update-field-options &match assoc ::outer-select true))
+
+    [:field id-or-name (opts :guard :join-alias)]
+    (let [{::add/keys [desired-alias]} (mbql.u/match-one (:source-query query)
+                                         [:field
+                                          (_ :guard (partial = id-or-name))
+                                          (matching-opts :guard #(= (:join-alias %) (:join-alias opts)))]
+                                         matching-opts)]
+      [:field id-or-name (cond-> opts
+                           desired-alias (assoc ::add/source-alias desired-alias
+                                                ::add/desired-alias desired-alias))])
+
+    ;; when recursing into joins use the refs from the parent level.
+    (m :guard (every-pred map? :joins))
+    (let [{:keys [joins]} m]
+      (-> (dissoc m :joins)
+          rewrite-fields-and-expressions
+          (assoc :joins (mapv (fn [join]
+                                (assoc join :qp/refs (:qp/refs query)))
+                              joins))))))
+
+(defn nest-expressions
+  "Pushes the `:source-table`/`:source-query`, `:expressions`, and `:joins` in the top-level of the query into a
```

Comment:
The calling bit says that it's idempotent, and it seems to be idempotent in first and second glance, but the only documentation that it should be idempotent seems to be at calling site a thousand miles away, so if we just edit this willy-nilly without being careful you wouldn't be able to continue guaranteeing idempotency - I think there should be some documentation here locally saying "keep this idempotent, folks" or something

Reply:
That's a great point